File: ValidateXmls.java
import java.io.File;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import javax.xml.XMLConstants;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.transform.stream.StreamSource;
import javax.xml.validation.Schema;
import javax.xml.validation.SchemaFactory;
import javax.xml.validation.Validator;
import org.w3c.dom.Document;
import org.w3c.dom.Element;

public class ValidateXmls {

    public static void main(String[] args) {
        String dataDirectoryPath = "/home/fernando/L2J_Mobius/L2J_Mobius_C4_ScionsOfDestiny/dist/game/data";
        if (args.length > 0) {
            dataDirectoryPath = args[0];
        }

        System.out.println("=== Iniciando Validación Masiva de XMLs contra XSD (Java) ===");
        int totalFiles = 0;
        int passed = 0;
        List<String[]> failed = new ArrayList<>();

        Path dataDir = Paths.get(dataDirectoryPath);

        try (Stream<Path> paths = Files.walk(dataDir)) {
            List<Path> xmlFiles = paths.filter(Files::isRegularFile)
                    .filter(p -> p.toString().endsWith(".xml"))
                    .collect(Collectors.toList());

            DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
            dbFactory.setNamespaceAware(true);
            DocumentBuilder dBuilder = dbFactory.newDocumentBuilder();
            SchemaFactory schemaFactory = SchemaFactory.newInstance(XMLConstants.W3C_XML_SCHEMA_NS_URI);

            for (Path xmlPath : xmlFiles) {
                totalFiles++;
                File xmlFile = xmlPath.toFile();
                try {
                    Document doc = dBuilder.parse(xmlFile);
                    Element rootElement = doc.getDocumentElement();
                    String xsdAttr = rootElement.getAttributeNS("http://www.w3.org/2001/XMLSchema-instance",
                            "noNamespaceSchemaLocation");

                    if (xsdAttr != null && !xsdAttr.isEmpty()) {
                        File xsdFile = new File(xmlFile.getParentFile(), xsdAttr);
                        if (!xsdFile.exists()) {
                            xsdFile = new File(dataDir.toFile(), xsdAttr);
                        }
                        if (!xsdFile.exists()) {
                            // Fallbacks
                            xsdFile = new File(new File(dataDir.toFile(), "xsd"), new File(xsdAttr).getName());
                        }

                        if (xsdFile.exists()) {
                            Schema schema = schemaFactory.newSchema(xsdFile);
                            Validator validator = schema.newValidator();
                            validator.validate(new StreamSource(xmlFile));
                            passed++;
                        } else {
                            failed.add(new String[] { xmlPath.toString(), "XSD no encontrado: " + xsdAttr });
                        }
                    } else {
                        failed.add(new String[] { xmlPath.toString(),
                                "No especifica XSD (xsi:noNamespaceSchemaLocation faltante)" });
                    }
                } catch (Exception e) {
                    failed.add(new String[] { xmlPath.toString(), "Error de parseo/validación: " + e.getMessage() });
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }

        System.out.println("\n=== Resumen de Validación ===");
        System.out.println("Total XML escaneados: " + totalFiles);
        System.out.println("Aprobados: " + passed);
        System.out.println("Con Errores o Sin XSD (" + failed.size() + "):");

        if (!failed.isEmpty()) {
            System.out.println("\n--- Detalles de Errores ---");
            for (String[] errorDetail : failed) {
                String relPath = dataDir.relativize(Paths.get(errorDetail[0])).toString();
                System.out.println("[X] " + relPath + " -> " + errorDetail[1]);
            }
        }
    }
}

File: validate_xmls.py
import os
import xml.etree.ElementTree as ET
try:
    from lxml import etree
    HAS_LXML = True
except ImportError:
    HAS_LXML = False

def validate_xml_files(data_dir):
    if not HAS_LXML:
        print("[Error] Se requiere 'lxml' para validar contra XSD. Instálalo con: pip install lxml")
        return

    print(f"=== Iniciando Validación Masiva de XMLs contra XSD ===")    
    total_files = 0
    passed = 0
    failed = []

    for root, _, files in os.walk(data_dir):
        for file in files:
            if file.endswith('.xml'):
                total_files += 1
                xml_path = os.path.join(root, file)
                try:
                    xml_doc = etree.parse(xml_path)
                    root_elem = xml_doc.getroot()
                    
                    # Extraer el atributo xsi:noNamespaceSchemaLocation
                    xsd_attr = root_elem.get('{http://www.w3.org/2001/XMLSchema-instance}noNamespaceSchemaLocation')
                    if xsd_attr:
                        # La ruta del xsd normalmente es algo como "xsd/SkillLearn.xsd" o "../../xsd/skills.xsd"
                        # Ajustar el path relativo desde el XML hasta el XSD
                        # En L2J Mobius, el esquema se resuelve asumiendo 'data/' como root, 
                        # pero resolveremos de forma conservadora.
                        xsd_path = os.path.normpath(os.path.join(data_dir, xsd_attr))
                        
                        if not os.path.exists(xsd_path):
                            # Fallback asumiendo rutas relativas o 'xsd/' dentro de data/
                            xsd_path_alt = os.path.join(data_dir, "xsd", os.path.basename(xsd_attr))
                            if os.path.exists(xsd_path_alt):
                                xsd_path = xsd_path_alt
                            else:
                                failed.append((xml_path, f"XSD no encontrado: {xsd_attr}"))
                                continue
                                
                        with open(xsd_path, 'r', encoding='utf-8') as f:
                            schema_doc = etree.parse(f)
                            schema = etree.XMLSchema(schema_doc)
                            
                        # Validar!
                        if schema.validate(xml_doc):
                            passed += 1
                        else:
                            failed.append((xml_path, str(schema.error_log.last_error)))
                    else:
                        failed.append((xml_path, "No especifica XSD (xsi:noNamespaceSchemaLocation faltante)"))
                except Exception as e:
                    failed.append((xml_path, f"Error de parseo: {str(e)}"))

    print("\n=== Resumen de Validación ===")
    print(f"Total XML escaneados: {total_files}")
    print(f"Aprobados: {passed}")
    print(f"Con Errores o Sin XSD: {len(failed)}")
    
    if failed:
        print("\n--- Detalles de Errores ---")
        for f, error in failed:
            rel_path = os.path.relpath(f, data_dir)
            print(f"[X] {rel_path} -> {error}")

if __name__ == '__main__':
    # Directorio hardcodeado a la ruta del Datapack
    data_directory = "/home/fernando/L2J_Mobius/L2J_Mobius_C4_ScionsOfDestiny/dist/game/data"
    validate_xml_files(data_directory)
